### PR TITLE
Update feature-flag-best-practices.md

### PR DESCRIPTION
**Pull request by riboflavin**
Updates to the Best Practices pages.

### CODE REVIEW

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 1]
```diff
@@ -1,339 +1,301 @@
...
 
-**5. Continuously Updated**: Implement update mechanisms in your application so that changes to feature flag configurations are reflected without requiring application restarts or redeployments. The SDK should handle subscriptions or polling to the feature flag service for updates.
+5\. Continuously Updated: Implement update mechanisms in your application so that changes to feature flag configurations can be made without requiring application restarts or redeployments. The SDK should handle subscriptions or polling to the feature flag service for updates.
 
-## 2. Never expose PII. Follow the principle of least privilege.
+<Figure caption="The SDK holds an in-memory feature flag configuration cache which is continuously synced with the feature flag control service. The SDK can then be used to check the state of feature flags in your application." img="/img/feature-flag-scalable-architecture.png"/>
+    
+### H3 \- 2\. Never expose PII. Follow the principle of least privilege.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#2-never-expose-pii-follow-the-principle-of-least-privilege)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:55]:
```suggestion
###\- 2\. Never expose PII. Follow the principle of least privilege.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#2-never-expose-pii-follow-the-principle-of-least-privilege)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 2]
```diff
@@ -1,339 +1,301 @@
...
+For client-side feature flags, you’ll need a different type of SDK. These SDKs will send the context to the Feature Flag Evaluation Service and receive the evaluated results. These results should be stored in memory for a feature flag lookup in the client-side application. By keeping the evaluated results for a specific context in memory in the client-side application, you avoid network roundtrips every time your application needs to check the status of a feature flag. It achieves the same level of performance as a server-side SDK, but the content stored in memory is different and limited to evaluated results on the client.
 
-2. **Cost Savings**:
+### H3 \- 3\. Evaluate flags as close to the user as possible. Reduce latency.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#3-evaluate-flags-as-close-to-the-user-as-possible-reduce-latency)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:89]:
```suggestion
### \- 3\. Evaluate flags as close to the user as possible. Reduce latency.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#3-evaluate-flags-as-close-to-the-user-as-possible-reduce-latency)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 3]
```diff
@@ -1,339 +1,301 @@
...
+4. Resilience:  
+   a. Service Outages: If the feature flag service experiences downtime or outages, local evaluation allows your application to continue functioning without interruptions. This is important for maintaining service reliability and ensuring your application remains available even when the service is down.
 
-In summary, this principle emphasizes the importance of optimizing performance while protecting end-user privacy by evaluating feature flags as close to the end user as possible. Done right, this also leads to a highly available feature flag system that scales with your applications.
+In summary, this principle emphasizes optimizing performance while protecting end-user privacy by evaluating feature flags as close to the end user as possible. This also leads to a highly available feature flag system that scales with your applications.
 
-## 4. Scale Horizontally. Decouple reading and writing flags.
+### H3 \- 4\. Scale Horizontally. Decouple reading and writing flags.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#4-scale-horizontally-decouple-reading-and-writing-flags)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:108]:
```suggestion
### \- 4\. Scale Horizontally. Decouple reading and writing flags.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#4-scale-horizontally-decouple-reading-and-writing-flags)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 4]
```diff
@@ -1,339 +1,301 @@
...
+6. Load Balancing:
 
-## 5. Limit payloads. Feature flag payload should be as small as possible.
+   * Load balancing strategies can be tailored to the specific needs of the read and write APIs. You can distribute traffic and resources accordingly to optimize performance and ensure that neither API becomes a bottleneck under heavy loads.
 
-Minimizing the size of feature flag payloads is a critical aspect of maintaining the efficiency and performance of a feature flag system. The configuration of your feature flags can vary in size depending on the complexity of your targeting rules. For instance, if you have a targeting engine that determines whether a feature flag should be active or inactive based on individual user IDs, you might be tempted to include all these user IDs within the configuration payload. While this approach may work fine for a small user base, it can become unwieldy when dealing with a large number of users.
+### H3 \- 5\. Limit payloads. Feature flag payload should be as small as possible.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#5-limit-payloads-feature-flag-payload-should-be-as-small-as-possible)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:140]:
```suggestion
### \- 5\. Limit payloads. Feature flag payload should be as small as possible.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#5-limit-payloads-feature-flag-payload-should-be-as-small-as-possible)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 5]
```diff
@@ -1,339 +1,301 @@
...
+   * A feature flagging system that consistently delivers small, manageable payloads is more likely to be reliable. It reduces the risk of network failures, timeouts, and other issues when handling large data transfers. Reliability is paramount for mission-critical applications.
 
-7. **Ease of Monitoring and Debugging**:
+7. Ease of Monitoring and Debugging:
 
-    - Smaller payloads are easier to monitor and debug. When issues arise, it's simpler to trace problems and identify their root causes when dealing with smaller, more manageable data sets.
+   * Smaller payloads are easier to monitor and debug. When issues arise, it's simpler to trace problems and identify their root causes when dealing with smaller, more manageable data sets.
 
-## 6. Design for failure. Favor availability over consistency.
+### H3 \- 6\. Design for failure. Favor availability over consistency.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#6-design-for-failure-favor-availability-over-consistency)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:174]:
```suggestion
### \- 6\. Design for failure. Favor availability over consistency.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#6-design-for-failure-favor-availability-over-consistency)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 6]
```diff
@@ -1,339 +1,301 @@
...
 
-5. **Documentation and Training:**
+   * Usage metrics: A feature flag system should provide insight on an aggregated level about the usage of feature flags. This is helpful for developers so that they can easily assess that everything works as expected.
 
-    - **Onboarding and Training:** The entire feature flag solution, including API, UI, and the SDKs, requires clear and comprehensive documentation, along with easy-to-understand examples, in order to simplify the onboarding process for new developers. It also supports the ongoing training of new team members, ensuring that everyone can effectively use the feature flagging solution.
+5. Documentation and Training:
 
-Thank you for reading
+   * Onboarding and Training: The entire feature flag solution, including API, UI, and SDKs, requires clear and comprehensive documentation, along with easy-to-understand examples, i order to simplify the onboarding process for new developers. It also supports the ongoing training of new team members, ensuring that everyone can effectively use the feature flagging solution.
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:301]:
```suggestion
   * Onboarding and Training: The entire feature flag solution, including API, UI, and SDKs, requires clear and comprehensive documentation, along with easy-to-understand examples, i order to simplify the onboarding process for new developers. It also supports the ongoing training of new team members, ensuring that everyone can effectively use the feature flagging solution.
   
Thank you for reading. Our motivation for writing these principles is to share what we’ve learned building a large-scale feature flag solution with other architects and engineers solving similar challenges. Unleash is open-source, and so are these principles. Have something to contribute? [Open a PR](https://github.com/Unleash/unleash/pulls) or [discussion](https://github.com/orgs/Unleash/discussions) on our Github.
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 7]
```diff
@@ -1,339 +1,301 @@
...
 
-By implementing these resilient architecture patterns, your feature flagging system can continue to function effectively even in the presence of downtime or network disruptions in the feature flagging service. This ensures that your main application remains stable, available, and resilient to potential issues in the feature flagging infrastructure, ultimately leading to a better user experience and improved reliability.
+### H3 \- 7\. Make feature flags short-lived. Do not confuse flags with application configuration.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#7-make-feature-flags-short-lived-do-not-confuse-flags-with-application-configuration)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:189]:
```suggestion
### \- 7\. Make feature flags short-lived. Do not confuse flags with application configuration.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#7-make-feature-flags-short-lived-do-not-confuse-flags-with-application-configuration)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 8]
```diff
@@ -1,339 +1,301 @@
...
+* Kill-switches \- these work like an inverted feature flag and are used to gracefully disable part of a system with known weak spots.
 
--   **Avoid zombies:** Uniqueness should be controlled using a global list of feature flag names. This prevents the reuse of old flag names to protect new features. Using old names can lead to accidental exposure of old features, still protected with the same feature flag name.
--   **Naming convention enforcement: **Ideally, unique names are enforced at creation time. In a large organization, it is impossible for all developers to know all flags used. Enforcing a naming convention makes naming easier, ensures consistency, and provides an easy way to check for uniqueness.
+* Internal flags are used to enable additional debugging, tracing, and metrics at runtime, which are too costly to run all the time. Software engineers can enable these while debugging issues.
 
-Unique naming has the following advantages:
+### H3 \- 8\. Use unique names across all applications. Enforce naming conventions.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#8-use-unique-names-across-all-applications-enforce-naming-conventions)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:209]:
```suggestion
### \- 8\. Use unique names across all applications. Enforce naming conventions.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#8-use-unique-names-across-all-applications-enforce-naming-conventions)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 9]
```diff
@@ -1,339 +1,301 @@
...
+* Easier to manage: When a flag has a unique name, it's easier to know what it is used for and where it is being used. For example, it will be easier to search across multiple code bases to find references for a feature flag when it has a unique identifier across the entire organization.
 
-    - **Correlating Changes with Metrics**: Engineers and product owners often need to correlate feature flag changes with production application metrics. This correlation helps them understand how feature flags affect user behavior, performance, and system health. It's essential for making data-driven decisions about feature rollouts, optimizations, or rollbacks.
+* Enables collaboration: A feature flag with a unique name in the organization, it simplifies collaboration across teams, products, and applications andensures that we all talk about the same feature.
 
-3. **Collaboration**:
+### H3 \- 9\. Choose open by default. Democratize feature flag access.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#9-choose-open-by-default-democratize-feature-flag-access)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:223]:
```suggestion
### \- 9\. Choose open by default. Democratize feature flag access.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#9-choose-open-by-default-democratize-feature-flag-access)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 10]
```diff
@@ -1,339 +1,301 @@
...
+5. Security and Compliance:
 
-4. **Support and Feedback**: Inconsistent experiences lead to confused users, increased support requests, and muddied user feedback. Consistency ensures that user issues are easier to identify and address.
+   * Security Audits: Users of a feature flag system should be part of corporate access control groups such as SSO. Sometimes, additional controls are necessary, such as feature flag approvals using the four-eyes principle.
 
-5. **Brand Integrity**: A consistent experience reflects positively on your brand. It demonstrates professionalism and commitment to user satisfaction, enhancing your brand's reputation.
+Access control and visibility into feature flag changes are essential for security and compliance purposes. It helps track and audit who has made changes to the system, which can be crucial in maintaining data integrity and adhering to regulatory requirements.
 
-**Strategies for Consistency in Percentage-Based Gradual Rollouts:**
+### H3 \- 10\. Do no harm. Prioritize consistent user experience.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#10-do-no-harm-prioritize-consistent-user-experience)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:253]:
```suggestion
### \- 10\. Do no harm. Prioritize consistent user experience.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#10-do-no-harm-prioritize-consistent-user-experience)
```

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 11]
```diff
@@ -1,339 +1,301 @@
...
 
-## 11. Enable traceability. Make it easy to understand flag evaluation.
+5. Transparent Communication: Clearly communicate the gradual rollout to users. Use in-app notifications, tooltips, or changelogs to inform users about changes, ensuring they know what to expect.
 
-Developer experience is a critical factor to consider when implementing a feature flag solution. A positive developer experience enhances the efficiency of the development process and contributes to the overall success and effectiveness of feature flagging. One crucial aspect of developer experience is ensuring the testability of the SDK and providing tools for developers to understand how and why feature flags are evaluated. This is important because:
+### H3 \- 11\. Enable traceability. Make it easy to understand flag evaluation.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#11-enable-traceability-make-it-easy-to-understand-flag-evaluation)
```
Comment (ferrantim) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:269]:
```suggestion
### \- 11\. Enable traceability. Make it easy to understand flag evaluation.[​](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#11-enable-traceability-make-it-easy-to-understand-flag-evaluation)
```

**Review by ferrantim — CHANGES_REQUESTED** [review]
Hi @riboflavin this looks generally good. I would like to add the images back and to fix the suggestions that I have made. Once you do that, you can resubmit the PR for approval.

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 12]
```diff
@@ -1,339 +1,301 @@
...
--   [6. Design for failure. Favor availability over consistency.](#6-design-for-failure-favor-availability-over-consistency)
--   [7. Make feature flags short-lived. Do not confuse flags with application configuration.](#7-make-feature-flags-short-lived-do-not-confuse-flags-with-application-configuration)
--   [8. Use unique names across all applications. Enforce naming conventions.](#8-use-unique-names-across-all-applications-enforce-naming-conventions)
--   [9. Choose open by default. Democratize feature flag access.](#9-choose-open-by-default-democratize-feature-flag-access)
--   [10. Do no harm. Prioritize consistent user experience.](#10-do-no-harm-prioritize-consistent-user-experience)
--   [11. Enable traceability. Make it easy to understand flag evaluation.](#11-enable-traceability-make-it-easy-to-understand-flag-evaluation)
+* [1\. Enable run-time control. Control flags dynamically, not using config files.](https://docs.getunleash.io/topics/feature-flags/feature-flag-best-practices\#1-enable-run-time-control-control-flags-dynamically-not-using-config-files)
```
Comment (melindafekete) [inline website/docs/topics/feature-flags/feature-flag-best-practices.md:15]: For consistency, I would go for runtime (unhyphenated) for all occurrences on the page

File: website/docs/topics/feature-flags/feature-flag-best-practices.md [hunk 13]
```diff
@@ -1,339 +1,301 @@
...
+* Local Cache: Maintaining a local cache of feature flag configuration helps reduce network round trips and dependency on external services. The local cache can be periodically synchronized with the central Feature Flag Control service when it's available. This approach minimizes the impact of network failures or service downtime on your application.
 
--   **Evaluate Locally**: Whenever possible, the SDKs or application components should be able to evaluate feature flags locally without relying on external services. This ensures that feature flag evaluations continue even when the feature flagging service is temporarily unavailable.
+* Evaluate Locally: Whenever possible, the SDKs or application components should be able to evaluate feature flags locally without relying on external services. This ensures that feature flag evaluations continue even when the feature flagging service is temporarily unavailable.
```
[the remark this answers is not in the document]
Comment (riboflavin) [reply website/docs/topics/feature-flags/feature-flag-best-practices.md]: I went through and tried to make these consistent.

**Comment by melindafekete** [comment]
Superseded by #7907

**Pull request closed without merging**